persistent_queue/persistent_queue.py

```diff
@@ -0,0 +1,464 @@
+            raise ValueError("filename must be set, e.g. '/my/path/important.queue'")
+        if maxsize < 0:
+            maxsize = 0
+
+        self.maxsize = maxsize
+        self.filename = os.path.abspath(filename)
+        self.dumps = dumps
+        self.loads = loads
+        self.flush_limit = flush_limit
+
+        self._file = self._open_file()
+        self._file_lock = threading.RLock()
+        self._get_lock = threading.RLock()
+        self._get_event = threading.Event()
+        self._put_lock = threading.RLock()
+        self._put_event = threading.Event()
+
+        self._all_tasks_done = threading.Condition()
+        self._unfinished_tasks = 0
+
+        self._file.seek(0, 0)
+        self._length = struct.unpack(HEADER_STRUCT[0], self._file.read(4))[0]
+
+    def _open_file(self, mode=None):
+        mode = mode or 'r+b' if os.path.isfile(self.filename) else 'w+b'
+        file = open(self.filename, mode=mode, buffering=0)
+
+        if mode == 'w+b':
+            # write length and start pointer
+            file.write(struct.pack(HEADER_STRUCT, 0, START_OFFSET))
+
+        return file
+
+    def _update_length(self, length):
+        current_pos = self._file.tell()
+
+        self._file.seek(0, 0)  # Go to the beginning of the file
+        self._file.write(struct.pack(HEADER_STRUCT[0], length))
+        self._file.flush()  # Probably not necessary since buffering=0
+        os.fsync(self._file.fileno())
+
+        self._length = length
+
+        self._file.seek(current_pos, 0)
+
+    def _get_queue_top(self):
+        current_pos = self._file.tell()
+
+        self._file.seek(START_OFFSET - 4, 0)  # Start at beginning of file
+        pos = struct.unpack(HEADER_STRUCT[1], self._file.read(4))[0]
+
+        self._file.seek(current_pos, 0)
+        return pos
+
+    def _set_queue_top(self, top):
+        current_pos = self._file.tell()
+
+        self._file.seek(START_OFFSET - 4, 0)  # Start at beginning of file
+        self._file.write(struct.pack(HEADER_STRUCT[1], top))
+        self._file.flush()  # Probably not necessary since buffering=0
+        os.fsync(self._file.fileno())
+
+        self._file.seek(current_pos, 0)
+
+    def _peek(self, block, timeout, items, partial=False):
+        """
+        Returns a certain amount of items from the queue. If items is greater
+        than one, a list is returned.
+        """
+        def read_data():
+            length = struct.unpack(LENGTH_STRUCT, self._file.read(4))[0]
+            data = self._file.read(length)
+            return self.loads(data)
+
+        _LOGGER.debug("Peeking %s items", items)
+
+        # Ignore requests for zero items
+        if items == 0:
+            _LOGGER.debug("Returning empty list")
+            return [], self._file.tell()
+
+        if block:
+            if timeout is not None:
+                target = time.time() + timeout
+            while self._length < items:
+                if self._put_event.wait(timeout) is not True:
```

Comment:
Can't 
```python
if self._put_event.wait(timeout) is not True:
``` 
be simplified to 
```python
if self._put_event.wait(timeout) is False:
```
?

Maybe I am understanding `Event.wait` incorrectly. Doesn't it return `False` if the `wait` call times out? Assuming that a `timeout` value is given, this branch

```python
if self._put_event.wait(timeout) is not True:
```
will only be taken if `wait` has timed out. Since `wait` has timed out, we can just `raise queue.Empty`, we don't need to do the rest of the logic you have. It is only when `wait` returns `True` (because something was added the the queue so the `wait` didn't time out) that you need to adjust `timeout` to a new value, as long as there's still not enough items in the queue.